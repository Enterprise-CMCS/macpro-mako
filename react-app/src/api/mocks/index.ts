export * as mockSubmit from "./submit";
export * as mockTypes from "./types";
