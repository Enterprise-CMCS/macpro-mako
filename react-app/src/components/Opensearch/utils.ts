import { opensearch } from "shared-types";

const filterMapQueryReducer = (
  state: Record<opensearch.Filterable<any>["prefix"], any[]>,
  filter: opensearch.Filterable<any>,
) => {
  if (filter.type === "exists") {
    state[filter.prefix].push({
      exists: { field: filter.field },
    });
  }

  if (filter.value === undefined || filter.value == null) return state;

  // this was hoisted up since false is a valid "match" value
  if (filter.type === "match") {
    state[filter.prefix].push({
      match: { [filter.field]: filter.value },
    });
  }

  if (!filter.value) return state;

  if (filter.type === "terms") {
    state[filter.prefix].push({
      terms: { [filter.field]: filter.value },
    });
  }

  if (filter.type === "term") {
    state[filter.prefix].push({
      term: { [filter.field]: filter.value },
    });
  }

  if (filter.type === "range") {
    state[filter.prefix].push({
      range: { [filter.field]: filter.value },
    });
  }

  if (filter.type === "global_search") {
    state[filter.prefix].push({
      dis_max: {
        tie_breaker: 0.7,
        boost: 1.2,
        queries: ["id", "submitterName", "leadAnalystName"].map((FIELD) => ({
          wildcard: {
            [`${FIELD}.keyword`]: {
              value: `*${(filter.value as string).trim()}*`,
              case_insensitive: true,
            },
          },
        })),
      },
    });
  }

  return state;
};

export const filterQueryBuilder = (filters: opensearch.Filterable<any>[]) => {
  if (!filters?.length) return {};

  return {
    query: {
      bool: filters.reduce(filterMapQueryReducer, {
        must: [],
        must_not: [],
        should: [],
        filter: [],
      }),
    },
  };
};

export const paginationQueryBuilder = (pagination: opensearch.QueryState<any>["pagination"]) => {
  if (!pagination) return {};

  const from = (() => {
    if (
      !pagination?.number ||
      !pagination?.size ||
      pagination?.number < 1 ||
      pagination?.size < 1
    ) {
      return 0;
    }
    return pagination.number * pagination.size;
  })();

  return {
    size: pagination?.size && pagination?.size > 0 ? pagination?.size : 25,
    from,
  };
};

export const sortQueryBuilder = (sort: opensearch.QueryState<any>["sort"]) => {
  if (!sort?.field) return {};
  return { sort: [{ [sort.field]: sort.order || "asc" }] };
};

export const aggQueryBuilder = (aggs: opensearch.AggQuery<any>[]) => {
  if (!aggs?.length) return {};

  return {
    aggs: aggs.reduce((STATE, AGG) => {
      if (AGG?.name && AGG?.type && AGG?.field) {
        STATE[AGG.name] = {
          [AGG.type]: {
            field: AGG.field,
            order: { _term: "asc" },
            ...(AGG.size && { size: AGG.size }),
          },
        };
      }
      return STATE;
    }, {} as any),
  };
};

export const createSearchFilterable = (value?: string) => {
  if (!value) return [];
  return [
    {
      type: "global_search",
      field: "",
      value,
      prefix: "must",
    } as unknown as opensearch.Filterable<any>,
  ];
};

export const checkMultiFilter = (filters: opensearch.Filterable<any>[] = [], val: number = 0) => {
  return (
    filters.length >= val ||
    filters.some((filter) => Array.isArray(filter.value) && filter.value.length >= val)
  );
};
