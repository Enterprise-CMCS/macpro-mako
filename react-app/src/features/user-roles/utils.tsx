export type StatusType = "active" | "pending" | "denied" | "revoked";
export type UserRoleType = {
  id: string;
  email: string;
  fullName: string;
  role: string;
  territory: string;
  doneByEmail: string;
  doneByName: string;
  lastModifiedDate: number;
  status: StatusType;
  eventType: string;
  group?: string;
  division?: string;
};

export const initSortUserData = (userData: UserRoleType[]) => {
  if (!userData.length) return [];
  // separate pending / other
  const pendingRoles = userData.filter((x: UserRoleType) => x.status === "pending");
  const remainingRoles = userData.filter((x: UserRoleType) => x.status !== "pending");

  const compare = (a: UserRoleType, b: UserRoleType) => {
    const nameA = a.fullName.toLocaleLowerCase();
    const nameB = b.fullName.toLocaleLowerCase();

    if (nameA < nameB) return -1;
    if (nameA > nameB) return 1;
    return 0;
  };
  const sorted = pendingRoles.sort(compare).concat(remainingRoles.sort(compare));

  return sorted;
};

export const sortUserData = (
  sortByKey: keyof UserRoleType,
  direction: boolean,
  data: UserRoleType[],
) => {
  if (!data.length) return [];

  // when direction is true, that means we are descending
  const [last, first] = direction ? [-1, 1] : [1, -1];
  const sortStatus = (a: UserRoleType, b: UserRoleType) => {
    switch (a.status) {
      case "pending":
        return first;
      case "active":
        return b.status === "pending" ? last : first;
      case "denied":
        return b.status === "revoked" ? first : last;
      case "revoked":
        return last;
      default:
        return last;
    }
  };
  const mainSort = (a: UserRoleType, b: UserRoleType) =>
    a[sortByKey] < b[sortByKey] ? first : last;

  if (sortByKey === "status") return data.sort((a, b) => sortStatus(a, b));

  return data.sort((a, b) => mainSort(a, b));
};
