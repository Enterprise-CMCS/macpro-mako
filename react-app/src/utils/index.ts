export * from "./createContextProvider";
export * from "./stateNames";
export * from "./textHelpers";
export * from "./labelMappers";
export * from "./user";
export * from "./crumbs";
export * from "./labels";
export * from "./utils";
export * from "./zod";
export * from "./formOrigin";
export * from "./location";
