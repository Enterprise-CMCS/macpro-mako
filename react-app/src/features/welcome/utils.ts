import { QueryClient } from "@tanstack/react-query";
import { redirect } from "react-router";

import { getUser } from "@/api";

export const loader = (queryClient: QueryClient) => {
  return async () => {
    const queryString = window.location.search;
    // Parse the query string to get URL parameters
    const queryParams = new URLSearchParams(queryString);
    // Access specific parameters
    const errorDescription = queryParams.get("error_description");
    const error = queryParams.get("error");
    if (errorDescription || error) {
      console.error("Authentication Error:", { errorDescription, error });
      return { error };
    }

    // check user query has been initialized
    if (!queryClient.getQueryData(["user"])) {
      const userFetch = await queryClient.fetchQuery({
        queryKey: ["user"],
        queryFn: () => getUser(),
      });
      const isUserLoggedIn =
        !["/login", "/faq"].includes(window.location.pathname) && !userFetch?.user;

      if (isUserLoggedIn) {
        return redirect("/login");
      }
    }

    return queryClient.getQueryData(["user"]);
  };
};
