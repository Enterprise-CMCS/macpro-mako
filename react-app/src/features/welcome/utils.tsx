import { QueryClient } from "@tanstack/react-query";
import { Navigate, redirect } from "react-router";

import { getUser, OneMacUser } from "@/api";

export const loader = (queryClient: QueryClient) => {
  return async () => {
    const queryString = window.location.search;
    // Parse the query string to get URL parameters
    const queryParams = new URLSearchParams(queryString);
    // Access specific parameters
    const errorDescription = queryParams.get("error_description");
    const error = queryParams.get("error");
    if (errorDescription || error) {
      console.error("Authentication Error:", { errorDescription, error });
      return { error };
    }

    // check user query has been initialized
    if (!queryClient.getQueryData(["user"])) {
      const userFetch = await queryClient.fetchQuery({
        queryKey: ["user"],
        queryFn: () => getUser(),
      });
      return !userFetch?.user && redirect("/login");
    }

    // check user is logged in
    const loginRedirect =
      !["/login", "/faq", "/support"].includes(window.location.pathname) &&
      !queryClient.getQueryData<OneMacUser>(["user"])?.user;
    if (loginRedirect) {
      return <Navigate to="/login" />;
    }

    return queryClient.getQueryData(["user"]);
  };
};
