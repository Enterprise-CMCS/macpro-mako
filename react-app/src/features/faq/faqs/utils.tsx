import { cn } from "@/utils";

export interface Template {
  title: string;
  text?: string;
  href: string;
  subtext?: string[];
}

export const helpDeskContact = {
  email: "OneMAC_Helpdesk@cms.hhs.gov",
  phone: "(833) 228-2540",
};

export const slugify = (text: string) =>
  text
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/(^-|-$)/g, "")
    .substring(0, 40);

export const handleSupportLinkClick = (type: string) => (e: React.MouseEvent<HTMLElement>) => {
  const text = e.currentTarget.textContent?.trim() || "unknown";
  window.gtag("event", `support_click_${type}_${slugify(text)}`, {
    event_category: "Support",
    event_label: text,
  });
};

export const PdfLink = ({
  href,
  title,
  text,
  label,
  className = "underline hover:no-underline",
}: {
  href: string;
  title: string;
  text?: string;
  label: string;
  className?: string;
}) => (
  <a
    className={cn("text-primary", className)}
    href={href}
    download
    rel="noopener noreferrer"
    target="_blank"
    onClick={handleSupportLinkClick(label)}
  >
    {title}
    {text && `: ${text}`}
  </a>
);

export const PdfList = ({
  list,
  label,
  ulClassName = "list-disc pl-6 space-y-2",
}: {
  list: Template[];
  label: string;
  ulClassName?: string;
}) => (
  <ul className={ulClassName} role="list">
    {list.map((pdf) => (
      <li key={pdf.title}>
        <PdfLink href={pdf.href} title={pdf.title} text={pdf.text} label={label} className="" />
        {pdf.subtext && (
          <ul className="list-disc pl-7 space-y-1" role="list">
            {pdf.subtext.map((sub, index) => (
              <li key={index} className="text-sm text-gray-600">
                {sub}
              </li>
            ))}
          </ul>
        )}
      </li>
    ))}
  </ul>
);
