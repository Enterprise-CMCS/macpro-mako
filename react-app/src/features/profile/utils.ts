import { UserRole } from "shared-types/events/legacy-user";

import { StateAccess } from "@/api";
import { convertStateAbbrToFullName } from "@/utils";

export const userRoleMap = {
  defaultcmsuser: "CMS Read-only User",
  cmsroleapprover: "CMS Role Approver",
  cmsreviewer: "CMS Read-only User",
  statesystemadmin: "State System Admin",
  helpdesk: "Helpdesk",
  statesubmitter: "State Submitter",
  systemadmin: "System Admin",
};

export const stateAccessRoles: UserRole[] = [
  "statesubmitter",
  "statesystemadmin",
  "cmsroleapprover",
  "systemadmin",
];

export const orderStateAccess = (accesses: StateAccess[]) => {
  if (!accesses || !accesses.length) return;
  // sort revoked states seprately and add to
  const activeStates = accesses.filter((x: StateAccess) => x.status != "revoked");
  const revokedStates = accesses.filter((x: StateAccess) => x.status == "revoked");

  const compare = (a: StateAccess, b: StateAccess) => {
    const stateA = convertStateAbbrToFullName(a.territory);
    const stateB = convertStateAbbrToFullName(b.territory);

    if (stateA < stateB) return -1;
    if (stateA > stateB) return 1;
    return 0;
  };

  const sorted = activeStates.sort(compare).concat(revokedStates.sort(compare));

  return sorted;
};

// if user has no active roles, show pending state(s)
// show state(s) for latest active role
export const filterStateAccess = (userDetails, userProfile) => {
  if (!userProfile?.stateAccess || userProfile.stateAccess.length < 1) return [];
  return userDetails?.role
    ? userProfile.stateAccess.filter(
        (access: StateAccess) => access.role === userDetails.role && access.territory !== "ZZ",
      )
    : userProfile.stateAccess.filter((access: StateAccess) => access.territory !== "ZZ");
};
