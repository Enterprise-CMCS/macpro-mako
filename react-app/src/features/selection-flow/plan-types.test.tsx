import { render, screen, waitForElementToBeRemoved } from "@testing-library/react";
import { setDefaultStateSubmitter, setMockUsername, testReviewer } from "mocks";
import React, { ReactElement } from "react";
import { beforeEach, describe, expect, it } from "vitest";

import { renderWithQueryClientAndMemoryRouter } from "@/utils/test-helpers";

import {
  B4WaiverSubmissionOptions,
  BCapWaiverSubmissionOptions,
  BWaiverSubmissionOptions,
  ChipSPASubmissionOptions,
  MedicaidSPASubmissionOptions,
  NewSubmissionInitialOptions,
  SimplePageTitle,
  SPASubmissionOptions,
  WaiverSubmissionOptions,
} from "./plan-types";

const setup = async (element: ReactElement, path: string) => {
  const rendered = renderWithQueryClientAndMemoryRouter(
    element,
    [
      {
        path: "/",
        element: <div>Home</div>,
      },
      {
        path,
        element,
      },
    ],
    {
      initialEntries: [
        {
          pathname: path,
        },
      ],
    },
  );
  if (screen.queryAllByLabelText("three-dots-loading")?.length > 0) {
    await waitForElementToBeRemoved(() => screen.queryAllByLabelText("three-dots-loading"));
  }
  return rendered;
};

describe("plan-types", () => {
  beforeEach(() => {
    setDefaultStateSubmitter();
  });

  describe("SimplePageTitle", () => {
    it("should render the page", () => {
      const { asFragment } = render(<SimplePageTitle title="Test" />);

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe("NewSubmissionInitialOptions", () => {
    it("should navigate unauthenticated users to /", async () => {
      setMockUsername(null);
      await setup(<NewSubmissionInitialOptions />, "/new-submission");
      expect(screen.getByText("Home")).toBeInTheDocument();
    });

    it("should navigate cms users to /", async () => {
      setMockUsername(testReviewer);
      await setup(<NewSubmissionInitialOptions />, "/new-submission");
      expect(screen.getByText("Home")).toBeInTheDocument();
    });

    it("should render the page for a state user", async () => {
      await setup(<NewSubmissionInitialOptions />, "/new-submission");
      expect(screen.getByRole("heading", { name: "Submission Type" })).toBeInTheDocument();
    });
  });

  describe("SPASubmissionOptions", () => {
    it("should render the page for a state user", async () => {
      await setup(<SPASubmissionOptions />, "/new-submission/spa");
      expect(screen.getByRole("heading", { name: "SPA Type" })).toBeInTheDocument();
    });
  });

  describe("MedicaidSPASubmissionOptions", () => {
    it("should render the page for a state user", async () => {
      await setup(<MedicaidSPASubmissionOptions />, "/new-submission/spa/medicaid");
      expect(screen.getByRole("heading", { name: "Medicaid SPA Type" })).toBeInTheDocument();
    });
  });

  describe("ChipSPASubmissionOptions", () => {
    it("should render the page for a state user", async () => {
      await setup(<ChipSPASubmissionOptions />, "/new-submission/spa/chip");
      expect(screen.getByRole("heading", { name: "CHIP SPA Type" })).toBeInTheDocument();
    });
  });

  describe("WaiverSubmissionOptions", () => {
    it("should render the page for a state user", async () => {
      await setup(<WaiverSubmissionOptions />, "/new-submission/waiver");
      expect(screen.getByRole("heading", { name: "Waiver Action Type" })).toBeInTheDocument();
    });
  });

  describe("BWaiverSubmissionOptions", () => {
    it("should render the page for a state user", async () => {
      await setup(<BWaiverSubmissionOptions />, "/new-submission/waiver/b");
      expect(
        screen.getByRole("heading", { name: "1915(b) Waiver Action Type" }),
      ).toBeInTheDocument();
    });
  });

  describe("B4WaiverSubmissionOptions", () => {
    it("should render the page for a state user", async () => {
      await setup(<B4WaiverSubmissionOptions />, "/new-submission/waiver/b/b4");
      expect(
        screen.getByRole("heading", {
          name: "1915(b)(4) FFS Selective Contracting Waiver Authority",
        }),
      ).toBeInTheDocument();
    });
  });

  describe("BCapWaiverSubmissionOptions", () => {
    it("should render the page for a state user", async () => {
      await setup(<BCapWaiverSubmissionOptions />, "/new-submission/waiver/b/capitated");
      expect(
        screen.getByRole("heading", { name: "1915(b) Comprehensive (Capitated) Waiver Authority" }),
      ).toBeInTheDocument();
    });
  });
});
