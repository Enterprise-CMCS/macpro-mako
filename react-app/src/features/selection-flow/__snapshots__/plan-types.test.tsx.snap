// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`plan-types > SimplePageTitle > should render the page 1`] = `
<DocumentFragment>
  <div
    class="flex items-center justify-between my-4"
  >
    <h1
      class="text-xl"
    >
      Test
    </h1>
  </div>
</DocumentFragment>
`;
